pathmux/viztree.go

```diff
@@ -0,0 +1,148 @@
+		}
+	}
+	numberOfChildren := len(tree.staticChild)
+	children := make([]*VizTree, 0, numberOfChildren)
+	for i := 0; i < numberOfChildren; i++ {
+		childrenNodes := aggregateTree(tree.staticChild[i], nextPath)
+		children = append(children, childrenNodes...)
+	}
+	childrenNodes := processWildCardNode(tree.wildcardChild)
+	// Keep lifting the leaf wildcard names for later processing
+	leafWildcardNames := getLeafNames(childrenNodes, currentNode != nil || middleNode != nil)
+	children = append(children, childrenNodes...)
+
+	if tree.catchAllChild != nil && tree.catchAllChild.leafValue != nil {
+		children = append(children, &VizTree{"*" + tree.catchAllChild.path, nil, true, sliceLeafNames(tree.catchAllChild.leafWildcardNames)})
+	}
+	if currentNode != nil {
+		if leafWildcardNames != nil {
+			currentNode.leafWildcardNames = leafWildcardNames
+		}
+		currentNode.children = children
+		return []*VizTree{currentNode}
+	}
+	if middleNode != nil {
+		// Prevent duplication of intermediate nodes
+		if previousNode := findNode(children, middleNode.Path); previousNode == nil {
+			if leafWildcardNames != nil {
+				middleNode.leafWildcardNames = leafWildcardNames
+			}
+			children = append(children, middleNode)
+		} else {
+			previousNode.canMatch = true
+			if leafWildcardNames != nil {
+				previousNode.leafWildcardNames = leafWildcardNames
+			}
+		}
+	}
+	return children
+}
+
+func processWildCardNode(tree *node) []*VizTree {
+	if tree == nil {
+		return nil
+	}
+	numberOfChildren := len(tree.staticChild)
+	if tree.leafWildcardNames != nil && numberOfChildren == 0 {
+		currentPath, leafWildcardNames := tree.leafWildcardNames[len(tree.leafWildcardNames)-1], sliceLeafNames(tree.leafWildcardNames)
+		return []*VizTree{{":" + currentPath, nil, tree.leafValue != nil, leafWildcardNames}}
+	}
+	children := make([]*VizTree, 0, numberOfChildren)
+	for i := 0; i < numberOfChildren; i++ {
+		childrenNodes := aggregateTree(tree.staticChild[i], "")
+		for j := 0; j < len(childrenNodes); j++ {
+			child := childrenNodes[j]
+			if child.leafWildcardNames != nil {
+				currentPath, leafWildcardNames := ":"+child.leafWildcardNames[len(child.leafWildcardNames)-1], sliceLeafNames(child.leafWildcardNames)
+				child.leafWildcardNames = nil
+				if previousNode := findNode(children, currentPath); previousNode == nil {
+					child = &VizTree{currentPath, []*VizTree{child}, tree.leafValue != nil, leafWildcardNames}
+				} else {
+					previousNode.children = append(previousNode.children, child)
+					child = nil
+				}
+			}
+			if child != nil {
+				children = append(children, child)
+			}
+		}
+	}
+
+	return children
+}
+
+func findNode(children []*VizTree, path string) *VizTree {
+	for i := 0; i < len(children); i++ {
+		child := children[i]
+		if path == child.Path {
+			return child
+		}
+	}
+	return nil
+}
+
+func getLeafNames(children []*VizTree, removeFromChildren bool) []string {
+	var leafWildcardNames []string
```

Comment:
this function is not completely clear to me. Should return the leafWildcardNames of the last child whose leafWildcardnames is not empty and also remove them if required?